var/spack/repos/builtin/packages/mpark-variant/icpc.patch

```diff
@@ -0,0 +1,33 @@
+From 2d933fe544bd5841e9016ab7e8066521ebe33f30 Mon Sep 17 00:00:00 2001
+From: sbolding <sbolding@lanl.gov>
+Date: Mon, 29 Mar 2021 19:13:28 -0600
+Subject: [PATCH] Apply patch for icpc
+
+icpc in some way utilizes the preprocessor of the associated "developer
+tools" used by the compiler. This leads to, in some cases, a
+preprocessor claiming support for `__tuple_element_packs`, even though
+icpc (as of version 21.1) can't actually parse such code.  Just use the
+MPARK_TUPLE_ELEMENT_PACK impl with __icc until icpc supports it.
+
+Fixes #77
+---
+ include/mpark/config.hpp | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/include/mpark/config.hpp b/include/mpark/config.hpp
+index f85ffb55c..128fa9235 100644
+--- a/include/mpark/config.hpp
++++ b/include/mpark/config.hpp
+@@ -50,7 +50,7 @@
+ #define MPARK_BUILTIN_UNREACHABLE
+ #endif
+ 
+-#if __has_builtin(__type_pack_element)
++#if __has_builtin(__type_pack_element) && !(defined(__ICC))
```

Comment:
Following this quick check on Linux, I think your problem is most likely macOS specific:
  https://godbolt.org/z/6v5GMxz4v

I would this recommend to here and in the upstream PR https://github.com/mpark/variant/pull/78 to limit yourself to ICC + Apple:
```suggestion
+#if __has_builtin(__type_pack_element) && !(defined(__ICC) && defined(__APPLE__))
```

Reply:
I originally thought this was an Apple specific issue, but I think it is just dependent on the toolchain icc is using (I may be saying this wrong; I really don't understand how Intel uses the underlying toolchain's preprocessor).  The reason this is passing on Linux is that with the gcc10.2 toolchain, `__has_builtin(__type_pack_element)` is evaluating to false: https://godbolt.org/z/zeevevqcW

My concern is that if in the future GCC adds `__type_pack_element` support, then Intel with that tool-chain will start failing.  It may also fail on linux if someone compiles using an LLVM toolchain: [intel doc on setting clang](https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/compiler-options/compiler-option-details/compatibility-options/clangxx-name.html.)  Other than compile time, I don't think there is much risk in always using mpark's implementation of type_pack_element.  

But I am open to maybe making it `__clang__` specific for now and dealing with it in the future when it arises :)

Reply:
I see, thank you for the details. Ok, works for me then!